core/embed/trezorhal/stm32f4/system.c

```diff
@@ -0,0 +1,315 @@
+
+      // --------------------------------------------------------------
+      // Clear .bss, initialize .data, ...
+      // --------------------------------------------------------------
+
+      "LDR     R0, =bss_start      \n"  // Clear .bss
+      "MOV     R1, #0              \n"
+      "LDR     R2, =bss_end        \n"
+      "SUB     R2, R2, R0          \n"
+      "BL      memset              \n"
+
+      "LDR     R0, =data_vma       \n"  // Initialize .data
+      "LDR     R1, =data_lma       \n"
+      "LDR     R2, =data_size      \n"
+      "BL      memcpy              \n"
+
+#ifdef STM32U5
+      "LDR     R0, =confidential_vma   \n"  // Initialize .confidental
+      "LDR     R1, =confidential_lma   \n"
+      "LDR     R2, =confidential_size  \n"
+      "BL      memcpy                  \n"
+#endif
+
+      // --------------------------------------------------------------
+      // Restore stack protector guard
+      // --------------------------------------------------------------
+
+      "LDR     R0, =%[STK_GUARD]   \n"  // Restore stack protector guard
+      "STR     R7, [R0]            \n"
+
+      // --------------------------------------------------------------
+      // Reset critical hardware so we can safely enable interrupts
+      // --------------------------------------------------------------
+
+      "BL      emergency_reset     \n"
+
+      "CPSIE   I                   \n"  // Re-enable interrupts
+
+      // --------------------------------------------------------------
+      // Clear all VFP registers
+      // --------------------------------------------------------------
+
+      "LDR     R1, = 0xE000EF34    \n"  // FPU->FPCCR
+      "LDR     R0, [R1]            \n"
+      "BIC     R0, R0, #1          \n"  // Clear LSPACT to suppress lazy
+                                        // stacking
+      "STR     R0, [R1]            \n"
+
+      // TODO: clear VFP registers (maybe for ARMV7-M only)
+
+      // --------------------------------------------------------------
+      // Clear R7-R11 registers
+      // --------------------------------------------------------------
+
+      "MOV     R0, #0              \n"
+      "MOV     R7, R0              \n"
+      "MOV     R8, R0              \n"
+      "MOV     R9, R0              \n"
+      "MOV     R10, R0             \n"
+      "MOV     R11, R0             \n"
+
+      // --------------------------------------------------------------
+      // Check if we are in thread mode and if yes, jump to error_handler
+      // --------------------------------------------------------------
+
+      "LDR      R1, =0x1FF         \n"  // Get lower 9 bits of IPSR
+      "MRS      R0, IPSR           \n"
+      "ANDS     R0, R0, R1         \n"
+      "CMP      R0, #0             \n"  // == 0 if in thread mode
+      "ITTT     EQ                 \n"
+      "MOVEQ    R0, R5             \n"  // R0 = pminfo
+      "LDREQ    LR, =reboot        \n"
```

Comment:
Yes, that's right. The handler should never return, but in case it does, `reboot` is called. Finally i think `shutdown_privileged` is more suitable (and consistent with error handler behaviour in systask.c). This is something that needs to be revisited later with T3W1.